From 11c31ff559000ee17a65fad2cd382471eccdb94a Mon Sep 17 00:00:00 2001
From: Adithya R <gh0strider.2k18.reborn@gmail.com>
Date: Mon, 7 Mar 2022 22:06:27 +0530
Subject: [PATCH 1/3] SystemUI: Add roaming indicator to statusbar tuner

Dhina17: Ported to 14
shutter-cat: adapt to A16 qpr0, qpr1
[@techyminati - Adapt to VollaOS 16 QPR2]

Change-Id: I326e1d54019179f0505a558b7e57ab6a2479cdf2
Co-authored-by: Dhina17 <dhinalogu@gmail.com>
Co-authored-by: Dmitrii <bankersenator@gmail.com>
Signed-off-by: Mohammad Hasan Keramat J <ikeramat@protonmail.com>
Signed-off-by: Pranav Vashi <neobuddy89@gmail.com>
Signed-off-by: Dmitrii <bankersenator@gmail.com>
---
 .../res/drawable/ic_statusbar_roaming.xml     | 25 +++++++++++++++++++
 packages/SystemUI/res/values/vos_strings.xml  | 10 ++++++++
 .../SystemUI/res/xml/status_bar_prefs.xml     |  5 ++++
 .../domain/interactor/MobileIconInteractor.kt |  4 +++
 .../interactor/MobileIconInteractorKairos.kt  |  4 +++
 .../MobileIconInteractorKairosAdapter.kt      |  9 ++++++-
 .../interactor/MobileIconsInteractor.kt       |  9 +++++++
 .../interactor/MobileIconsInteractorKairos.kt | 10 ++++++++
 .../MobileIconsInteractorKairosAdapter.kt     |  6 +++++
 .../mobile/ui/binder/MobileIconBinder.kt      |  2 +-
 .../ui/viewmodel/MobileIconViewModel.kt       | 17 +++++++++++++
 .../viewmodel/MobileIconsViewModelKairos.kt   |  1 +
 .../shared/data/model/ConnectivitySlots.kt    |  5 +++-
 13 files changed, 104 insertions(+), 3 deletions(-)
 create mode 100644 packages/SystemUI/res/drawable/ic_statusbar_roaming.xml
 create mode 100644 packages/SystemUI/res/values/vos_strings.xml

diff --git a/packages/SystemUI/res/drawable/ic_statusbar_roaming.xml b/packages/SystemUI/res/drawable/ic_statusbar_roaming.xml
new file mode 100644
index 000000000000..b8d2be6543ab
--- /dev/null
+++ b/packages/SystemUI/res/drawable/ic_statusbar_roaming.xml
@@ -0,0 +1,25 @@
+<!--
+     Copyright (C) 2021 The Android Open Source Project
+
+     Licensed under the Apache License, Version 2.0 (the "License");
+     you may not use this file except in compliance with the License.
+     You may obtain a copy of the License at
+
+          http://www.apache.org/licenses/LICENSE-2.0
+
+     Unless required by applicable law or agreed to in writing, software
+     distributed under the License is distributed on an "AS IS" BASIS,
+     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+     See the License for the specific language governing permissions and
+     limitations under the License.
+-->
+<vector xmlns:android="http://schemas.android.com/apk/res/android"
+        android:width="16dp"
+        android:height="16dp"
+        android:viewportWidth="16.0"
+        android:viewportHeight="16.0"
+        android:tint="?android:attr/colorControlNormal">
+    <path
+        android:fillColor="#FFFFFFFF"
+        android:pathData="M4.5,13.02V3h3.93c0.97,0 1.75,0.27 2.34,0.81c0.6,0.53 0.9,1.28 0.9,2.25c0,0.63 -0.19,1.2 -0.56,1.74c-0.37,0.53 -0.95,0.91 -1.74,1.12l2.45,4.02v0.08h-1.69L7.75,9.09H6.03v3.93H4.5zM8.46,4.44H6.03v3.23h2.44c0.49,0 0.9,-0.14 1.2,-0.41c0.31,-0.28 0.46,-0.69 0.46,-1.22s-0.15,-0.93 -0.45,-1.2C9.38,4.58 8.98,4.44 8.46,4.44z"/>
+</vector>
diff --git a/packages/SystemUI/res/values/vos_strings.xml b/packages/SystemUI/res/values/vos_strings.xml
new file mode 100644
index 000000000000..196e110a17e6
--- /dev/null
+++ b/packages/SystemUI/res/values/vos_strings.xml
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+     SPDX-FileCopyrightText: 2026 Volla Systeme GmbH
+     SPDX-License-Identifier: Apache-2.0
+-->
+<resources xmlns:xliff="urn:oasis:names:tc:xliff:document:1.2">
+
+    <!-- Name of the roaming status bar icon. -->
+    <string name="status_bar_roaming">Roaming indicator</string>
+</resources>
diff --git a/packages/SystemUI/res/xml/status_bar_prefs.xml b/packages/SystemUI/res/xml/status_bar_prefs.xml
index 0c6a850b7fe9..bc242ecc1f71 100644
--- a/packages/SystemUI/res/xml/status_bar_prefs.xml
+++ b/packages/SystemUI/res/xml/status_bar_prefs.xml
@@ -101,6 +101,11 @@
         android:key="airplane"
         android:title="@string/status_bar_airplane" />
 
+    <com.android.systemui.tuner.StatusBarSwitch
+        android:icon="@drawable/ic_statusbar_roaming"
+        android:key="roaming"
+        android:title="@string/status_bar_roaming" />
+
     <!-- other weird signal stuff -->
 
     <com.android.systemui.tuner.StatusBarSwitch
diff --git a/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/mobile/domain/interactor/MobileIconInteractor.kt b/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/mobile/domain/interactor/MobileIconInteractor.kt
index a49214461d11..4c607ea6f4d8 100644
--- a/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/mobile/domain/interactor/MobileIconInteractor.kt
+++ b/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/mobile/domain/interactor/MobileIconInteractor.kt
@@ -124,6 +124,9 @@ interface MobileIconInteractor {
      */
     val isRoaming: Flow<Boolean>
 
+    /** See [MobileIconsInteractor.isRoamingForceHidden]. */
+    val isRoamingForceHidden: Flow<Boolean>
+
     /** See [MobileIconsInteractor.isForceHidden]. */
     val isForceHidden: Flow<Boolean>
 
@@ -147,6 +150,7 @@ class MobileIconInteractorImpl(
     defaultMobileIconGroup: StateFlow<MobileIconGroup>,
     isDefaultConnectionFailed: StateFlow<Boolean>,
     override val isForceHidden: Flow<Boolean>,
+    override val isRoamingForceHidden: Flow<Boolean>,
     connectionRepository: MobileConnectionRepository,
     private val context: Context,
     val carrierIdOverrides: MobileIconCarrierIdOverrides = MobileIconCarrierIdOverridesImpl(),
diff --git a/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/mobile/domain/interactor/MobileIconInteractorKairos.kt b/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/mobile/domain/interactor/MobileIconInteractorKairos.kt
index f9be9e49de55..c53799b8c632 100644
--- a/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/mobile/domain/interactor/MobileIconInteractorKairos.kt
+++ b/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/mobile/domain/interactor/MobileIconInteractorKairos.kt
@@ -124,6 +124,9 @@ interface MobileIconInteractorKairos {
     /** See [MobileIconsInteractor.isForceHidden]. */
     val isForceHidden: State<Boolean>
 
+    /** See [MobileIconsInteractor.isRoamingForceHidden]. */
+    val isRoamingForceHidden: State<Boolean>
+
     /** See [MobileConnectionRepository.isAllowedDuringAirplaneMode]. */
     val isAllowedDuringAirplaneMode: State<Boolean>
 
@@ -143,6 +146,7 @@ class MobileIconInteractorKairosImpl(
     defaultMobileIconGroup: State<MobileIconGroup>,
     isDefaultConnectionFailed: State<Boolean>,
     override val isForceHidden: State<Boolean>,
+    override val isRoamingForceHidden: State<Boolean>,
     private val connectionRepository: MobileConnectionRepositoryKairos,
     private val context: Context,
     private val carrierIdOverrides: MobileIconCarrierIdOverrides =
diff --git a/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/mobile/domain/interactor/MobileIconInteractorKairosAdapter.kt b/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/mobile/domain/interactor/MobileIconInteractorKairosAdapter.kt
index c725cc92478e..99dcd38312ab 100644
--- a/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/mobile/domain/interactor/MobileIconInteractorKairosAdapter.kt
+++ b/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/mobile/domain/interactor/MobileIconInteractorKairosAdapter.kt
@@ -131,6 +131,13 @@ fun BuildScope.MobileIconInteractorKairosAdapter(
                         "MobileIconInteractorKairosAdapter(subId=$subscriptionId).isForceHidden"
                     },
                 ),
+            isRoamingForceHidden =
+                isRoamingForceHidden.toColdConflatedFlow(
+                    kairosNetwork,
+                    nameTag {
+                        "MobileIconInteractorKairosAdapter(subId=$subscriptionId).isRoamingForceHidden"
+                    },
+                ),
             isAllowedDuringAirplaneMode =
                 isAllowedDuringAirplaneMode.toStateFlow(
                     nameTag {
@@ -145,7 +152,6 @@ fun BuildScope.MobileIconInteractorKairosAdapter(
                 ),
         )
     }
-
 private class MobileIconInteractorKairosAdapter(
     override val subscriptionId: Int,
     override val tableLogBuffer: TableLogBuffer,
@@ -165,6 +171,7 @@ private class MobileIconInteractorKairosAdapter(
     override val isSingleCarrier: StateFlow<Boolean>,
     override val isRoaming: StateFlow<Boolean>,
     override val isForceHidden: Flow<Boolean>,
+    override val isRoamingForceHidden: Flow<Boolean>,
     override val isAllowedDuringAirplaneMode: StateFlow<Boolean>,
     override val carrierNetworkChangeActive: StateFlow<Boolean>,
 ) : MobileIconInteractor
diff --git a/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/mobile/domain/interactor/MobileIconsInteractor.kt b/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/mobile/domain/interactor/MobileIconsInteractor.kt
index e468c0556326..9b1a0fc80823 100644
--- a/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/mobile/domain/interactor/MobileIconsInteractor.kt
+++ b/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/mobile/domain/interactor/MobileIconsInteractor.kt
@@ -125,6 +125,9 @@ interface MobileIconsInteractor {
     /** True if we're configured to force-hide the mobile icons and false otherwise. */
     val isForceHidden: Flow<Boolean>
 
+    /** True if we're configured to force-hide the roaming and false otherwise. */
+    val isRoamingForceHidden: Flow<Boolean>
+
     /**
      * True if the device-level service state (with -1 subscription id) reports emergency calls
      * only. This value is only useful when there are no other subscriptions OR all existing
@@ -425,6 +428,11 @@ constructor(
             .map { it.contains(ConnectivitySlot.MOBILE) }
             .stateIn(scope, SharingStarted.WhileSubscribed(), false)
 
+    override val isRoamingForceHidden: Flow<Boolean> =
+        connectivityRepository.forceHiddenSlots
+            .map { it.contains(ConnectivitySlot.ROAMING) }
+            .stateIn(scope, SharingStarted.WhileSubscribed(), false)
+
     override val isDeviceInEmergencyCallsOnlyMode: Flow<Boolean> =
         mobileConnectionsRepo.isDeviceEmergencyCallCapable
 
@@ -444,6 +452,7 @@ constructor(
                 defaultMobileIconGroup,
                 isDefaultConnectionFailed,
                 isForceHidden,
+                isRoamingForceHidden,
                 mobileConnectionsRepo.getRepoForSubId(subId),
                 context,
             )
diff --git a/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/mobile/domain/interactor/MobileIconsInteractorKairos.kt b/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/mobile/domain/interactor/MobileIconsInteractorKairos.kt
index c3385abe0214..baf393e96a58 100644
--- a/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/mobile/domain/interactor/MobileIconsInteractorKairos.kt
+++ b/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/mobile/domain/interactor/MobileIconsInteractorKairos.kt
@@ -124,6 +124,9 @@ interface MobileIconsInteractorKairos {
     /** True if we're configured to force-hide the mobile icons and false otherwise. */
     val isForceHidden: State<Boolean>
 
+    /** True if we're configured to force-hide the roaming icon and false otherwise. */
+    val isRoamingForceHidden: State<Boolean>
+
     /**
      * True if the device-level service state (with -1 subscription id) reports emergency calls
      * only. This value is only useful when there are no other subscriptions OR all existing
@@ -429,6 +432,12 @@ constructor(
             .map { it.contains(ConnectivitySlot.MOBILE) }
     }
 
+    override val isRoamingForceHidden: State<Boolean> = buildState {
+        connectivityRepository.forceHiddenSlots.toState().map {
+            it.contains(ConnectivitySlot.ROAMING)
+        }
+    }
+
     override val isDeviceInEmergencyCallsOnlyMode: State<Boolean>
         get() = mobileConnectionsRepo.isDeviceEmergencyCallCapable
 
@@ -446,6 +455,7 @@ constructor(
             defaultMobileIconGroup,
             isDefaultConnectionFailed,
             isForceHidden,
+	    isRoamingForceHidden,
             repo,
             context,
         )
diff --git a/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/mobile/domain/interactor/MobileIconsInteractorKairosAdapter.kt b/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/mobile/domain/interactor/MobileIconsInteractorKairosAdapter.kt
index e22fb922250b..111c076b1989 100644
--- a/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/mobile/domain/interactor/MobileIconsInteractorKairosAdapter.kt
+++ b/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/mobile/domain/interactor/MobileIconsInteractorKairosAdapter.kt
@@ -123,6 +123,11 @@ constructor(
             )
             .stateIn(scope, SharingStarted.WhileSubscribed(), false)
 
+    override val isRoamingForceHidden: Flow<Boolean> =
+        kairosInteractor.isRoamingForceHidden
+            .toColdConflatedFlow(kairosNetwork)
+            .stateIn(scope, SharingStarted.WhileSubscribed(), false)
+
     override val activeMobileDataSubscriptionId: StateFlow<Int?>
         get() = repo.activeMobileDataSubscriptionId
 
@@ -233,6 +238,7 @@ constructor(
             override val isSingleCarrier: Flow<Boolean> = latest { isSingleCarrier }
             override val isRoaming: Flow<Boolean> = latest { isRoaming }
             override val isForceHidden: Flow<Boolean> = latest { isForceHidden }
+            override val isRoamingForceHidden: Flow<Boolean> = latest { isRoamingForceHidden }
             override val isAllowedDuringAirplaneMode: Flow<Boolean> = latest {
                 isAllowedDuringAirplaneMode
             }
diff --git a/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/mobile/ui/binder/MobileIconBinder.kt b/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/mobile/ui/binder/MobileIconBinder.kt
index 87c128a8dd0e..05e7bb02ab12 100644
--- a/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/mobile/ui/binder/MobileIconBinder.kt
+++ b/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/mobile/ui/binder/MobileIconBinder.kt
@@ -225,7 +225,7 @@ object MobileIconBinder {
 
                     // Set the roaming indicator
                     launch {
-                        viewModel.roaming.distinctUntilChanged().collect { isRoaming ->
+                        viewModel.isRoamingVisible.distinctUntilChanged().collect { isRoaming ->
                             if (NewStatusBarIcons.isEnabled) {
                                 endSideRoamingView.isVisible = isRoaming
                             } else {
diff --git a/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/mobile/ui/viewmodel/MobileIconViewModel.kt b/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/mobile/ui/viewmodel/MobileIconViewModel.kt
index 3b40cdec8721..df905e0ce806 100644
--- a/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/mobile/ui/viewmodel/MobileIconViewModel.kt
+++ b/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/mobile/ui/viewmodel/MobileIconViewModel.kt
@@ -50,6 +50,7 @@ interface MobileIconViewModelCommon {
     val icon: Flow<SignalIconModel>
     val contentDescription: Flow<MobileContentDescription?>
     val roaming: Flow<Boolean>
+    val isRoamingVisible: Flow<Boolean>
     /** The RAT icon (LTE, 3G, 5G, etc) to be displayed. Null if we shouldn't show anything */
     val networkTypeIcon: Flow<Icon.Resource?>
     /** The slice attribution. Drawn as a background layer */
@@ -125,6 +126,9 @@ class MobileIconViewModel(
 
     override val roaming: Flow<Boolean> = vmProvider.flatMapLatest { it.roaming }
 
+    override val isRoamingVisible: Flow<Boolean> =
+        vmProvider.flatMapLatest { it.isRoamingVisible }
+
     override val networkTypeIcon: Flow<Icon.Resource?> =
         vmProvider.flatMapLatest { it.networkTypeIcon }
 
@@ -161,6 +165,7 @@ private class CarrierBasedSatelliteViewModelImpl(
 
     /** These fields are not used for satellite icons currently */
     override val roaming: Flow<Boolean> = flowOf(false)
+    override val isRoamingVisible: Flow<Boolean> = flowOf(false)
     override val networkTypeIcon: Flow<Icon.Resource?> = flowOf(null)
     override val networkTypeBackground: StateFlow<Icon.Resource?> = MutableStateFlow(null)
     override val activityInVisible: Flow<Boolean> = flowOf(false)
@@ -301,6 +306,18 @@ private class CellularIconViewModel(
             )
             .stateIn(scope, SharingStarted.WhileSubscribed(), false)
 
+    override val isRoamingVisible: StateFlow<Boolean> =
+        combine(
+                roaming,
+                iconInteractor.isRoamingForceHidden
+            ) { isRoaming, isHidden ->
+                // If it's force hidden, just hide.
+                // Otherwise follow roaming state
+                isRoaming && !isHidden
+            }
+            .distinctUntilChanged()
+            .stateIn(scope, SharingStarted.WhileSubscribed(), false)
+
     private val activity: Flow<DataActivityModel?> =
         if (!constants.shouldShowActivityConfig) {
             flowOf(null)
diff --git a/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/mobile/ui/viewmodel/MobileIconsViewModelKairos.kt b/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/mobile/ui/viewmodel/MobileIconsViewModelKairos.kt
index 0457fbbbd231..4184f80b2889 100644
--- a/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/mobile/ui/viewmodel/MobileIconsViewModelKairos.kt
+++ b/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/mobile/ui/viewmodel/MobileIconsViewModelKairos.kt
@@ -165,6 +165,7 @@ constructor(
             override val isSingleCarrier: State<Boolean> = latest(true) { isSingleCarrier }
             override val isRoaming: State<Boolean> = latest(false) { isRoaming }
             override val isForceHidden: State<Boolean> = latest(false) { isForceHidden }
+            override val isRoamingForceHidden: State<Boolean> = latest(false) { isRoamingForceHidden }
             override val isAllowedDuringAirplaneMode: State<Boolean> =
                 latest(false) { isAllowedDuringAirplaneMode }
             override val carrierNetworkChangeActive: State<Boolean> =
diff --git a/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/shared/data/model/ConnectivitySlots.kt b/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/shared/data/model/ConnectivitySlots.kt
index d52d0fb91b82..9850784d6608 100644
--- a/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/shared/data/model/ConnectivitySlots.kt
+++ b/packages/SystemUI/src/com/android/systemui/statusbar/pipeline/shared/data/model/ConnectivitySlots.kt
@@ -30,12 +30,14 @@ class ConnectivitySlots @Inject constructor(context: Context) {
     private val mobileSlot: String = context.getString(R.string.status_bar_mobile)
     private val wifiSlot: String = context.getString(R.string.status_bar_wifi)
     private val ethernetSlot: String = context.getString(R.string.status_bar_ethernet)
+    private val roamingSlot: String = "roaming"
 
     private val slotByName: Map<String, ConnectivitySlot> = mapOf(
         airplaneSlot to ConnectivitySlot.AIRPLANE,
         mobileSlot to ConnectivitySlot.MOBILE,
         wifiSlot to ConnectivitySlot.WIFI,
-        ethernetSlot to ConnectivitySlot.ETHERNET
+        ethernetSlot to ConnectivitySlot.ETHERNET,
+        roamingSlot to ConnectivitySlot.ROAMING
     )
 
     /**
@@ -52,5 +54,6 @@ enum class ConnectivitySlot {
     AIRPLANE,
     ETHERNET,
     MOBILE,
+    ROAMING,
     WIFI,
 }
-- 
2.34.1

